### PR TITLE
Time dependent boundary conditions for auxiliary physics

### PR DESCRIPTION
**Pull request by oguevremont**
# Description of the problem

- Auxiliary physics didn't allow to update boundary conditions in time, unlike the flow solver
- Time dependent parameters (boolean) for auxiliary physics were not initialized

# Description of the solution

- A new function to update independently any auxiliary physics was added, to be overridden in the specific classes
- An mechanism to update the boundary conditions only for necessary physics was added  

# How Has This Been Tested?

- All tests still pass
- [lethe-fluid/time_dependent_boundaries_heat_transfer]
- [lethe-fluid/time_dependent_boundaries_tracer]
- [lethe-fluid/time_dependent_boundaries_vof]
- [lethe-fluid/time_dependent_boundaries_ch]

# Documentation

- The boundary conditions parameters pages are updated

# Future changes

- We could add space dependent parameters for convection-radiation, but for the moment it did not seem necessary

# Comments

- I misnamed the branch initially

### CODE REVIEW

File: doc/source/parameters/cfd/boundary_conditions_cfd.rst [hunk 1]
```diff
@@ -72,7 +72,7 @@ where `beta` is a constant and  :math:`(\mathbf{u}\cdot n)_{-}` is :math:`min (0
 
     Likewise, if ``number = 2`` and there is no ``subsection bc 0`` explicitly stated, the boundary is still created, with ``none`` by default.
 
-* ``time dependent`` specifies if a  boundary condition is time dependent (``true``) or steady (``false```). By default, this parameter is set to ``false``. This is there to improve the computational efficiency for transient cases in which the boundary conditions do not change. 
+* ``time dependent`` specifies if a  boundary condition is time dependent (``true``) or steady (``false``). By default, this parameter is set to ``false``. This is here to improve the computational efficiency for transient cases in which the boundary conditions do not change.
```
Comment (blaisb) [inline doc/source/parameters/cfd/boundary_conditions_cfd.rst:75]:
```suggestion
* ``time dependent`` specifies if a  boundary condition is time dependent (``true``) or steady (``false``). By default, this parameter is set to ``false``. This improves the computational efficiency for transient cases in which the boundary conditions do not change.
```

File: doc/source/parameters/cfd/boundary_conditions_multiphysics.rst [hunk 2]
```diff
@@ -15,23 +15,35 @@ The default parameters for ``temperature`` and ``convection-radiation`` are show
...
+        set Function expression = 0
+      end
     end
     set Stefan-Boltzmann constant = 0.000000056703
   end
 
-* ``number``: This is the number of boundary conditions of the problem. 
+* ``number``: This is the number of boundary conditions of the problem.
+
+* ``time dependent`` specifies if a boundary condition is time dependent (``true``) or steady (``false``). By default, this parameter is set to ``false``. This is here to improve the computational efficiency for transient cases in which the boundary conditions do not change.
```
Comment (blaisb) [inline doc/source/parameters/cfd/boundary_conditions_multiphysics.rst:46]:
```suggestion
* ``time dependent`` specifies if a boundary condition is time dependent (``true``) or steady (``false``). By default, this parameter is set to ``false``. This improves the computational efficiency for transient cases in which the boundary conditions do not change.
```

File: doc/source/parameters/cfd/boundary_conditions_multiphysics.rst [hunk 3]
```diff
@@ -46,7 +58,7 @@ The default parameters for ``temperature`` and ``convection-radiation`` are show
 * ``type``: type of boundary condition being imposed. At the moment, choices are:
     * ``noflux`` (default) so that there is no heat transfer boundary condition,
     * ``temperature`` (Dirichlet BC), to impose a given temperature ``value`` at the boundary,
-    * ``convection-radiation`` (Robin BC) for cooling/heating, depending on the environment temperature at the boundary ``Tinf``, with a given heat transfer coefficient ``h`` and emissivity of the boundary :math:`\mathbf{\epsilon}` following Newton's law of cooling (and heating) and Stefan-Boltzmann law of radiation.
+    * ``convection-radiation`` (Robin BC) for cooling/heating, depending on the environment temperature at the boundary ``Tinf``, with a given heat transfer coefficient ``h`` and emissivity of the boundary :math:`\mathbf{\epsilon}` following Newton's law of cooling (and heating) and Stefan-Boltzmann law of radiation. Note that the expressions for ``h``, ``Tinf`` and ``emissivity`` can be time dependent, but the current implementation doesn't allow for space dependence (the expressions are evaluated at the origin).
```
Comment (blaisb) [inline doc/source/parameters/cfd/boundary_conditions_multiphysics.rst:61]: This could be fixed in the future quite easily I believe

File: include/core/boundary_conditions.h [hunk 4]
```diff
@@ -440,11 +440,11 @@ namespace BoundaryConditions
   class HTBoundaryConditions : public BoundaryConditions<dim>
   {
   public:
-    std::vector<double> value;
-    std::vector<double> h;
-    std::vector<double> Tinf;
-    std::vector<double> emissivity;
-    double              Stefan_Boltzmann_constant;
+    std::vector<std::shared_ptr<Functions::ParsedFunction<dim>>> value;
```
Comment (blaisb) [inline include/core/boundary_conditions.h:443]: maybe the value should be renamed dirichlet_value

File: source/solvers/heat_transfer.cc [hunk 5]
```diff
@@ -1499,10 +1557,12 @@ HeatTransfer<dim>::postprocess_heat_flux_on_bc(
                       // Gather h coefficient and T_inf
                       const double h_coefficient =
                         this->simulation_parameters.boundary_conditions_ht
-                          .h[vector_index];
+                          .h[vector_index]
```
Comment (blaisb) [inline source/solvers/heat_transfer.cc:1560]: I will take care of generalizing this within the Hackathon I think

File: source/solvers/tracer.cc [hunk 6]
```diff
@@ -712,6 +712,44 @@ Tracer<dim>::setup_dofs()
...
+              ->set_time(time);
+            VectorTools::interpolate_boundary_values(
+              this->dof_handler,
+              this->simulation_parameters.boundary_conditions_tracer.id[i_bc],
+              *this->simulation_parameters.boundary_conditions_tracer
+                 .tracer[i_bc],
+              nonzero_constraints);
+          }
+      }
+  }
```
Comment (blaisb) [inline source/solvers/tracer.cc:746]: Why is there this additional scope here, it's not necessary no?

File: source/solvers/heat_transfer.cc [hunk 7]
```diff
@@ -1082,6 +1081,65 @@ HeatTransfer<dim>::setup_dofs()
...
+            BoundaryConditions::BoundaryType::temperature)
+          {
+            VectorTools::interpolate_boundary_values(
+              this->dof_handler,
+              this->simulation_parameters.boundary_conditions_ht.id[i_bc],
+              *this->simulation_parameters.boundary_conditions_ht.value[i_bc],
+              nonzero_constraints);
+          }
+      }
+  }
```
Comment (blaisb) [inline source/solvers/heat_transfer.cc:1139]: Why do you have this additional scope here?

**Review by blaisb — CHANGES_REQUESTED** [review]
Very minor changes, very cool PR man.

File: doc/source/parameters/cfd/boundary_conditions_multiphysics.rst (hunk 3 above)
Comment (AmishgaAlphonius) [inline doc/source/parameters/cfd/boundary_conditions_multiphysics.rst:61]:
```suggestion
    * ``convection-radiation`` (Robin BC) for cooling/heating, depending on the environment temperature at the boundary ``Tinf``, with a given heat transfer coefficient ``h`` and emissivity of the boundary :math:`\mathbf{\epsilon}` following Newton's law of cooling (and heating) and Stefan-Boltzmann law of radiation. Note that the expressions for ``h``, ``Tinf`` and ``emissivity`` can be time-dependent, but the current implementation doesn't allow for space dependence (the expressions are evaluated at the origin).
```

File: doc/source/parameters/cfd/boundary_conditions_cfd.rst (hunk 1 above)
Comment (AmishgaAlphonius) [reply doc/source/parameters/cfd/boundary_conditions_cfd.rst:75]:
```suggestion
* ``time dependent`` specifies if a  boundary condition is time-dependent (``true``) or steady (``false``). By default, this parameter is set to ``false``. This improves the computational efficiency for transient cases in which the boundary conditions do not change.
```

File: doc/source/parameters/cfd/boundary_conditions_multiphysics.rst (hunk 2 above)
Comment (AmishgaAlphonius) [reply doc/source/parameters/cfd/boundary_conditions_multiphysics.rst:46]:
```suggestion
* ``time dependent`` specifies if a boundary condition is time-dependent (``true``) or steady (``false``). By default, this parameter is set to ``false``. This improves the computational efficiency for transient cases in which the boundary conditions do not change.
```

File: doc/source/parameters/cfd/boundary_conditions_multiphysics.rst [hunk 8]
```diff
@@ -79,6 +92,8 @@ For tracer boundary conditions, the defaults parameters are:
 
 * ``number``: This is the number of boundary conditions of the problem. 
 
+* ``time dependent`` specifies if a boundary condition is time dependent (``true``) or steady (``false``). By default, this parameter is set to ``false``. This is here to improve the computational efficiency for transient cases in which the boundary conditions do not change.
```
Comment (AmishgaAlphonius) [inline doc/source/parameters/cfd/boundary_conditions_multiphysics.rst:95]:
```suggestion
* ``time dependent`` specifies if a boundary condition is time-dependent (``true``) or steady (``false``). By default, this parameter is set to ``false``. This improves the computational efficiency for transient cases in which the boundary conditions do not change.
```

File: doc/source/parameters/cfd/boundary_conditions_multiphysics.rst [hunk 9]
```diff
@@ -114,7 +130,9 @@ For VOF boundary conditions (multiphase flow), the possible ``types`` are ``none
 
     Likewise, if ``number = 2`` and there is no ``subsection bc 0`` explicitly stated, the boundary is still created, with ``none`` by default.
 
-* ``number``: This is the number of boundary conditions of the problem. 
+* ``number``: This is the number of boundary conditions of the problem.
+
+* ``time dependent`` specifies if a boundary condition is time dependent (``true``) or steady (``false``). By default, this parameter is set to ``false``. This is here to improve the computational efficiency for transient cases in which the boundary conditions do not change.
```
Comment (AmishgaAlphonius) [inline doc/source/parameters/cfd/boundary_conditions_multiphysics.rst:135]:
```suggestion
* ``time dependent`` specifies if a boundary condition is time-dependent (``true``) or steady (``false``). By default, this parameter is set to ``false``. This improves the computational efficiency for transient cases in which the boundary conditions do not change.
```

File: include/core/boundary_conditions.h [hunk 10]
```diff
@@ -521,8 +523,19 @@ namespace BoundaryConditions
                         "0",
                         Patterns::Integer(),
                         "Number of boundary conditions");
+
+      prm.declare_entry(
+        "time dependent",
+        "false",
+        Patterns::Bool(),
+        "Bool to define if the boundary condition is time dependent");
```
Comment (AmishgaAlphonius) [inline include/core/boundary_conditions.h:531]:
```suggestion
        "Bool to define if the boundary condition is time-dependent");
```

File: include/core/boundary_conditions.h [hunk 11]
```diff
@@ -560,20 +573,30 @@ namespace BoundaryConditions
...
       {
-        this->type[i_bc]       = BoundaryType::convection_radiation;
-        this->h[i_bc]          = prm.get_double("h");
-        this->Tinf[i_bc]       = prm.get_double("Tinf");
-        this->emissivity[i_bc] = prm.get_double("emissivity");
+        this->type[i_bc] = BoundaryType::convection_radiation;
 
-        Assert(this->emissivity[i_bc] <= 1.0 && this->emissivity[i_bc] >= 0.0,
-               EmissivityError(this->emissivity[i_bc]));
+        // Emissivity validity (0 <= emissivity <=1) will be checked at
```
Comment (AmishgaAlphonius) [inline include/core/boundary_conditions.h:582]:
```suggestion
        // Emissivity validity (0 <= emissivity <= 1) will be checked at
```

File: include/core/boundary_conditions.h [hunk 12]
```diff
@@ -689,6 +713,13 @@ namespace BoundaryConditions
                         "0",
                         Patterns::Integer(),
                         "Number of boundary conditions");
+
+      prm.declare_entry(
+        "time dependent",
+        "false",
+        Patterns::Bool(),
+        "Bool to define if the boundary condition is time dependent");
```
Comment (AmishgaAlphonius) [inline include/core/boundary_conditions.h:721]:
```suggestion
        "Bool to define if the boundary condition is time-dependent");
```

File: include/core/boundary_conditions.h [hunk 13]
```diff
@@ -1008,6 +1039,11 @@ namespace BoundaryConditions
                         "0",
                         Patterns::Integer(),
                         "Number of boundary conditions");
+      prm.declare_entry(
+        "time dependent",
+        "false",
+        Patterns::Bool(),
+        "Bool to define if the boundary condition is time dependent");
```
Comment (AmishgaAlphonius) [inline include/core/boundary_conditions.h:1046]:
```suggestion
        "Bool to define if the boundary condition is time-dependent");
```

File: include/solvers/auxiliary_physics.h [hunk 14]
```diff
@@ -103,6 +103,12 @@ class AuxiliaryPhysics : public PhysicsSolver<VectorType>
   virtual void
   modify_solution(){};
 
+  /**
+   * @brief Update non zero constraints if the boundary is time dependent
```
Comment (AmishgaAlphonius) [inline include/solvers/auxiliary_physics.h:107]:
```suggestion
   * @brief Update non zero constraints if the boundary is time-dependent
```

File: include/solvers/cahn_hilliard.h [hunk 15]
```diff
@@ -250,6 +250,12 @@ class CahnHilliard : public AuxiliaryPhysics<dim, TrilinosWrappers::MPI::Vector>
   void
   set_initial_conditions() override;
 
+  /**
+   * @brief Update non zero constraints if the boundary is time dependent
```
Comment (AmishgaAlphonius) [inline include/solvers/cahn_hilliard.h:254]:
```suggestion
   * @brief Update non zero constraints if the boundary is time-dependent
```

File: include/solvers/gls_navier_stokes.h [hunk 16]
```diff
@@ -65,7 +65,7 @@ class GLSNavierStokesSolver
   setup_dofs_fd() override;
 
   /**
-   * @brief update non zero constraint if the boundary is time dependant
+   * @brief update non zero constraint if the boundary is time dependent
```
Comment (AmishgaAlphonius) [inline include/solvers/gls_navier_stokes.h:68]:
```suggestion
   * @brief update non zero constraint if the boundary is time-dependent
```

File: include/solvers/heat_transfer.h [hunk 17]
```diff
@@ -233,6 +233,12 @@ class HeatTransfer : public AuxiliaryPhysics<dim, TrilinosWrappers::MPI::Vector>
   void
   set_initial_conditions() override;
 
+  /**
+   * @brief Update non zero constraints if the boundary is time dependent
```
Comment (AmishgaAlphonius) [inline include/solvers/heat_transfer.h:237]:
```suggestion
   * @brief Update non zero constraints if the boundary is time-dependent
```

File: include/solvers/multiphysics_interface.h [hunk 18]
```diff
@@ -267,6 +267,22 @@ class MultiphysicsInterface
       }
   }
 
+  /**
+   * @param Update the boundary conditions of the auxiliary physics if they are time dependent
```
Comment (AmishgaAlphonius) [inline include/solvers/multiphysics_interface.h:271]:
```suggestion
   * @param Update the boundary conditions of the auxiliary physics if they are time-dependent
```

File: include/solvers/tracer.h [hunk 19]
```diff
@@ -217,6 +217,12 @@ class Tracer : public AuxiliaryPhysics<dim, TrilinosWrappers::MPI::Vector>
   void
   set_initial_conditions() override;
 
+  /**
+   * @brief Update non zero constraints if the boundary is time dependent
```
Comment (AmishgaAlphonius) [inline include/solvers/tracer.h:221]:
```suggestion
   * @brief Update non zero constraints if the boundary is time-dependent
```

File: source/solvers/gls_navier_stokes.cc [hunk 20]
```diff
@@ -191,25 +191,29 @@ template <int dim>
 void
 GLSNavierStokesSolver<dim>::update_boundary_conditions()
 {
-  double time = this->simulation_control->get_current_time();
-  for (unsigned int i_bc = 0;
-       i_bc < this->simulation_parameters.boundary_conditions.size;
-       ++i_bc)
+  if (this->simulation_parameters.boundary_conditions.time_dependent)
```
Comment (AmishgaAlphonius) [inline source/solvers/gls_navier_stokes.cc:194]: Is this better (more efficient) than checking if false + return like in the auxiliary physics or is it the same?

**Review by AmishgaAlphonius — CHANGES_REQUESTED** [review]
Great addition to the code! :D
I have a few comments, here and there, but the most important, I think, is to update parameters in examples also.

File: doc/source/parameters/cfd/boundary_conditions_multiphysics.rst [hunk 21]
```diff
@@ -15,23 +15,35 @@ The default parameters for ``temperature`` and ``convection-radiation`` are show
 .. code-block:: text
 
   subsection boundary conditions heat transfer
-    set number = 2
+    set number         = 2
+    set time dependent = false
     subsection bc 0
       set id    = 0
       set type  = temperature
       set value = 0
```
[the remark this answers is not in the document]
Comment (oguevremont) [reply doc/source/parameters/cfd/boundary_conditions_multiphysics.rst:23]: Good catch! I forgot, but I'll fix it now

File: source/solvers/gls_navier_stokes.cc (hunk 20 above)
Comment (oguevremont) [reply source/solvers/gls_navier_stokes.cc:194]: Good question. I checked and there is no performance difference, but `if(boolean) return` is more easy to read and is more explicit, so I'll put it like that. I'll make it uniform though.

File: source/solvers/heat_transfer.cc (hunk 7 above)
Comment (oguevremont) [reply source/solvers/heat_transfer.cc:1139]: I put it because that's the way that it was written in the setup_dofs functions, but without this additional scope it works just as well so I'll remove it.

File: source/solvers/tracer.cc (hunk 6 above)
Comment (oguevremont) [reply source/solvers/tracer.cc:746]: I put it because that's the way that it was written in the setup_dofs functions, but without this additional scope it works just as well so I'll remove it.

File: include/core/boundary_conditions.h (hunk 4 above)
Comment (oguevremont) [reply include/core/boundary_conditions.h:443]: I agree. Did you mean only in the code or even the name of the parameter?

Comment (blaisb) [reply include/core/boundary_conditions.h]: in the code I think would be sufficient.

File: source/solvers/gls_navier_stokes.cc (hunk 20 above)
Comment (blaisb) [reply source/solvers/gls_navier_stokes.cc]: here speed is of no concern TBH

File: doc/source/parameters/cfd/boundary_conditions_multiphysics.rst (hunk 3 above)
Comment (oguevremont) [reply doc/source/parameters/cfd/boundary_conditions_multiphysics.rst]: Would it be ok to do the functions evaluation at the face barycenter, and add the value to the HT scratch data for the assembly? I can do it now if you'd like

Comment (blaisb) [reply doc/source/parameters/cfd/boundary_conditions_multiphysics.rst]: Well in this case I would calculate it at the face gauss point location and add it to the scratch data. it would be that simple really. Just like we do with the Nitsche weak immersed boundary in NS.

**Review by blaisb — APPROVED** [review]

File: doc/source/examples/multiphysics/warming-up-a-viscous-fluid/warming-up-a-viscous-fluid.rst
Comment (AmishgaAlphonius) [inline doc/source/examples/multiphysics/warming-up-a-viscous-fluid/warming-up-a-viscous-fluid.rst:1]: Lines 360, 361, and 366 need to be changed also :)

File: doc/source/parameters/cfd/boundary_conditions_cfd.rst (hunk 1 above)
Comment (AmishgaAlphonius) [inline doc/source/parameters/cfd/boundary_conditions_cfd.rst:75]:
```suggestion
* ``time dependent`` specifies if a  boundary condition is time-dependent (``true``) or steady (``false``). By default, this parameter is set to ``false``. This is here to improve the computational efficiency for transient cases in which the boundary conditions do not change.
```

File: doc/source/parameters/cfd/boundary_conditions_multiphysics.rst (hunk 21 above)
Comment (AmishgaAlphonius) [inline doc/source/parameters/cfd/boundary_conditions_multiphysics.rst:23]:
```suggestion
```

File: doc/source/parameters/cfd/boundary_conditions_multiphysics.rst (hunk 2 above)
Comment (AmishgaAlphonius) [inline doc/source/parameters/cfd/boundary_conditions_multiphysics.rst:46]:
```suggestion
* ``time dependent`` specifies if a boundary condition is time-dependent (``true``) or steady (``false``). By default, this parameter is set to ``false``. This is here to improve the computational efficiency for transient cases in which the boundary conditions do not change.
```

File: doc/source/parameters/cfd/boundary_conditions_multiphysics.rst
Comment (AmishgaAlphonius) [inline doc/source/parameters/cfd/boundary_conditions_multiphysics.rst:1]: Line 178: change "boundary condition is **time dependent** (``true``)" for "boundary condition is **time-dependent** (``true``)" (without the bold format)

**Review by AmishgaAlphonius — COMMENTED** [review]
Last few minor comments on my side. :)

File: doc/source/parameters/cfd/boundary_conditions_cfd.rst (hunk 1 above)
Comment (PierreLaurentinCS) [inline doc/source/parameters/cfd/boundary_conditions_cfd.rst:75]:
```suggestion
* ``time dependent`` specifies if a boundary condition is time dependent (``true``) or steady (``false``). By default, this parameter is set to ``false``. This is here to improve the computational efficiency for transient cases in which the boundary conditions do not change.
```

File: source/solvers/cahn_hilliard.cc [hunk 22]
```diff
@@ -917,6 +917,55 @@ CahnHilliard<dim>::setup_dofs()
...
+            this->simulation_parameters.boundary_conditions_cahn_hilliard
+              .id[i_bc],
+            CahnHilliardFunctionDefined<dim>(
+              &this->simulation_parameters.boundary_conditions_cahn_hilliard
+                 .bcFunctions[i_bc]
+                 .phi),
+            nonzero_constraints,
+            mask);
+        }
+    }
```
Comment (PierreLaurentinCS) [inline source/solvers/cahn_hilliard.cc:962]:
```suggestion
    }
    
```

**Review by PierreLaurentinCS — COMMENTED** [review]
Very good addition to the code!

**Review by AmishgaAlphonius — APPROVED** [review]
